src/protocol/securemul.rs

```diff
@@ -14,6 +15,48 @@ pub struct DValue<F> {
     d: F,
 }
 
+/// `SecureMul` for vector multiplications
+#[derive(Debug)]
+pub struct SecureMulAll<'a, N> {
+    prss: &'a PrssSpace,
+    gateway: &'a Gateway<N>,
+    step: &'a UniqueStepId,
+}
+
+impl<'a, N: Network> SecureMulAll<'a, N> {
+    pub fn new(prss: &'a PrssSpace, gateway: &'a Gateway<N>, step: &'a UniqueStepId) -> Self {
+        Self {
+            prss,
+            gateway,
+            step,
+        }
+    }
+
+    /// Executes the secure multiplications on the given slices. This is a wrapper
+    /// on `SecureMul` that iteratively performs multiplications on elements in the
+    /// given lists positioned at the same index.
+    ///
+    /// Each multiplication is assigned a `RecordId` starting from 0. Calling this
+    /// method multiple times in parts could result in a bad state.
+    pub async fn execute<F: Field>(
+        self,
+        a: &[Replicated<F>],
+        b: &[Replicated<F>],
```

Comment:
I chose slices because the input vectors lengths are finite, and don't need to take the ownership of them. I could do `Iterator<Item = &Replicated<F>>` (note the &) then the caller can call this with `iter()` instead of `into_iter()`.

If the input vectors are generated with the streaming multiplication, then iterators would be more appropriate I think.

Reply:
Unfortunately, I'm too much of Rust n00b to know what we should do here. This seems like a find place to start and we can always modify it later if we need / want to.

Reply:
So when you say finite, do you mean finite in the sense that it is always a small number that is independent of the number of records?  In that case, provided that the number of records that the inputs span (I know that you are processing in increasingly large jumps) is sufficiently small (number: TBD) then a slice is fine.  On the other hand, if it could be a very large span of the total set of records or dependent on the number of records, an iterator is better.

A stream would be even better, I suspect, but then you would also be producing a stream.